File: fsharp/AoC2021/Day07/Solution.fs
module Solution

let diff (a: uint) (b: uint) = if a > b then a - b else b - a

let computeFuelNaive (points: seq<uint>) (point: uint) =
    points
    |> Seq.map (fun p -> diff p point)
    |> Seq.sum

let computeFuelCorrect (points: seq<uint>) (point: uint) =
    points
    |> Seq.map (fun p -> diff p point)
    |> Seq.map (fun d -> d * (d + 1u) / 2u)
    |> Seq.sum

let findOptimalPoint (computeFuel: seq<uint> -> uint -> uint) (points: seq<uint>) =
    let min = points |> Seq.min
    let max = points |> Seq.max
    let cost = computeFuel points

    { min .. max }
    |> Seq.map (fun point -> (point, cost point))
    |> Seq.minBy snd

module Tests =
    open FsUnit
    open Xunit

    [<Fact>]
    let ``Test naive fuel cost`` () =
        ([ 16u
           1u
           2u
           0u
           4u
           2u
           7u
           1u
           2u
           14u ],
         2u)
        ||> computeFuelNaive
        |> should equal 37u

    [<Fact>]
    let ``Test correct fuel cost`` () =
        ([ 16u
           1u
           2u
           0u
           4u
           2u
           7u
           1u
           2u
           14u ],
         2u)
        ||> computeFuelCorrect
        |> should equal 206u

        ([ 16u
           1u
           2u
           0u
           4u
           2u
           7u
           1u
           2u
           14u ],
         5u)
        ||> computeFuelCorrect
        |> should equal 168u

    let ``Test finding the optimal point`` () =
        [ 16u
          1u
          2u
          0u
          4u
          2u
          7u
          1u
          2u
          14u ]
        |> findOptimalPoint computeFuelNaive
        |> should equal (2u, 37u)

        [ 16u
          1u
          2u
          0u
          4u
          2u
          7u
          1u
          2u
          14u ]
        |> findOptimalPoint computeFuelCorrect
        |> should equal (5u, 168u)

File: csharp/AoC2021/Day07/Solution.cs
namespace Day07;

public static class Solution
{
    public static (uint, uint)? FindOptimalPoint(List<uint> points,
        Func<IEnumerable<uint>, uint, uint> computeFuelCost)
    {
        var min = uint.MaxValue;
        var max = uint.MinValue;
        foreach (var point in points)
        {
            if (point < min) min = point;
            if (point > max) max = point;
        }

        (uint, uint)? optimum = null;
        for (var point = min; point <= max; point += 1)
        {
            var fuelCost = computeFuelCost(points, point);
            optimum = optimum switch
            {
                null => (point, fuelCost),
                var (_, c) when c > fuelCost => (point, fuelCost),
                _ => optimum
            };
        }

        return optimum;
    }

    public static uint ComputeFuelCostNaive(IEnumerable<uint> points, uint point) =>
        points
            .Select(p => Diff(p, point))
            .Aggregate((acc, p) => acc + p);

    public static uint ComputeFuelCostCorrect(IEnumerable<uint> points, uint point) =>
        points
            .Select(p => Diff(p, point))
            .Select(d => d * (d + 1) / 2)
            .Aggregate((acc, p) => acc + p);

    private static uint Diff(uint p1, uint p2) => p1 > p2 ? p1 - p2 : p2 - p1;
}
